File: ios/Sources/RynBridgeAnalytics/AnalyticsModule.swift
import Foundation
import RynBridge

public struct AnalyticsModule: BridgeModule, Sendable {
    public let name = "analytics"
    public let version = "0.1.0"
    public let actions: [String: ActionHandler]

    public init(provider: AnalyticsProvider) {
        actions = [
            "logEvent": { payload in
                guard let eventName = payload["name"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: name")
                }
                let params: [String: AnyCodable]
                if let dict = payload["params"]?.dictionaryValue {
                    params = dict
                } else {
                    params = [:]
                }
                provider.logEvent(name: eventName, params: params)
                return [:]
            },
            "setUserProperty": { payload in
                guard let key = payload["key"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: key")
                }
                guard let value = payload["value"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: value")
                }
                provider.setUserProperty(key: key, value: value)
                return [:]
            },
            "setUserId": { payload in
                guard let userId = payload["userId"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: userId")
                }
                provider.setUserId(userId)
                return [:]
            },
            "setScreen": { payload in
                guard let screenName = payload["name"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: name")
                }
                provider.setScreen(name: screenName)
                return [:]
            },
            "resetUser": { _ in
                provider.resetUser()
                return [:]
            },
            "setEnabled": { payload in
                guard let enabled = payload["enabled"]?.boolValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: enabled")
                }
                let result = try await provider.setEnabled(enabled)
                return ["enabled": .bool(result)]
            },
            "isEnabled": { _ in
                let result = try await provider.isEnabled()
                return ["enabled": .bool(result)]
            },
        ]
    }
}

File: android/analytics/src/main/kotlin/io/rynbridge/analytics/AnalyticsModule.kt
package io.rynbridge.analytics

import io.rynbridge.core.*

class AnalyticsModule(provider: AnalyticsProvider) : BridgeModule {
    constructor() : this(DefaultAnalyticsProvider())

    override val name = "analytics"
    override val version = "0.1.0"
    override val actions: Map<String, ActionHandler> = mapOf(
        "logEvent" to { payload ->
            val eventName = payload["name"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: name")
            val params: Map<String, BridgeValue> = payload["params"]?.dictionaryValue ?: emptyMap()
            provider.logEvent(eventName, params)
            emptyMap()
        },
        "setUserProperty" to { payload ->
            val key = payload["key"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: key")
            val value = payload["value"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: value")
            provider.setUserProperty(key, value)
            emptyMap()
        },
        "setUserId" to { payload ->
            val userId = payload["userId"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: userId")
            provider.setUserId(userId)
            emptyMap()
        },
        "setScreen" to { payload ->
            val screenName = payload["name"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: name")
            provider.setScreen(screenName)
            emptyMap()
        },
        "resetUser" to { _ ->
            provider.resetUser()
            emptyMap()
        },
        "setEnabled" to { payload ->
            val enabled = payload["enabled"]?.boolValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: enabled")
            val result = provider.setEnabled(enabled)
            mapOf("enabled" to BridgeValue.bool(result))
        },
        "isEnabled" to { _ ->
            val result = provider.isEnabled()
            mapOf("enabled" to BridgeValue.bool(result))
        }
    )
}
